### PR TITLE
Fix numpy d410

### PR DESCRIPTION
**Pull request by farmersez**

### CODE REVIEW

File: src/pydocstyle/checker.py [hunk 1]
```diff
@@ -34,6 +39,16 @@ def decorator(f):
     return decorator
 
 
+def pairwise(iterable, default_value):
```
Comment (Nurdok) [inline src/pydocstyle/checker.py]: Move this to `utils.py`.

File: src/pydocstyle/checker.py [hunk 2]
```diff
@@ -486,8 +501,8 @@ def _check_section_underline(cls, section_name, context, indentation):
 
         if not dash_line_found:
             yield violations.D407(section_name)
-            if next_non_empty_line_offset == 0:
-                yield violations.D410(section_name)
+            if next_non_empty_line_offset > 0:
```
Comment (Nurdok) [inline src/pydocstyle/checker.py]: `next_non_empty_line_offset` is a confusing name. How about something like `blank_lines_after_header`?

File: src/pydocstyle/checker.py [hunk 3]
```diff
@@ -498,22 +513,24 @@ def _check_section_underline(cls, section_name, context, indentation):
...
 
-            line_after_dashes = \
-                context.following_lines[next_non_empty_line_offset + 1]
-            if not is_blank(line_after_dashes):
-                yield violations.D410(section_name)
-
             if leading_space(dash_line) > indentation:
                 yield violations.D215(section_name)
 
+            if next_non_empty_line_offset + 1 < len(context.following_lines):
```
Comment (Nurdok) [inline src/pydocstyle/checker.py]: If I understand correctly, this means "if this is not the last section", right? Please add a comment to explain this.

File: src/pydocstyle/violations.py [hunk 4]
```diff
@@ -223,6 +223,8 @@ def to_rst(cls):
                          'Expected {0!r} dashes in section {1!r}, got {2!r}')
 D410 = D4xx.create_error('D410', 'Missing blank line after section', '{0!r}')
 D411 = D4xx.create_error('D411', 'Missing blank line before section', '{0!r}')
+D412 = D4xx.create_error('D412', 'Section content should be in the line '
```
Comment (Nurdok) [inline src/pydocstyle/violations.py]: I suggest making the description here clearer: "No blank lines allowed between a section header and its content".

File: src/tests/test_cases/sections.py [hunk 5]
```diff
@@ -75,7 +90,7 @@ def no_blank_line_after_section():
 
     Returns
     -------
-    A whole lot of values.
+    A value of some sort.
```
Comment (Nurdok) [inline src/tests/test_cases/sections.py]: I think the "blank line after a section" should not be enforced for the last section of the docstring (as discussed in #230).

File: src/pydocstyle/violations.py (hunk 4 above)
Comment (farmersez) [inline src/pydocstyle/violations.py]: 👍

File: src/pydocstyle/checker.py (hunk 1 above)
Comment (farmersez) [inline src/pydocstyle/checker.py]: 👍

File: src/pydocstyle/checker.py (hunk 2 above)
Comment (farmersez) [inline src/pydocstyle/checker.py]: Legit :^)

File: src/pydocstyle/checker.py (hunk 3 above)
Comment (farmersez) [inline src/pydocstyle/checker.py]: 👍

Comment (farmersez) [inline src/pydocstyle/checker.py]: I changed some of the logic here. The method is a bit long, but I don't think there's a non-artificial way of splitting it to several ones. Tell me if you think I should split it anyways.

File: src/tests/test_cases/sections.py (hunk 5 above)
Comment (Nurdok) [inline src/tests/test_cases/sections.py]: @farmersez what did you end up doing?

Comment (farmersez) [inline src/tests/test_cases/sections.py]: I split it to a different error code which is not raised by default in the `numpy` convention.

**Pull request merged**